contracts/exchange/OneSidedExchange.sol

```diff
@@ -0,0 +1,186 @@
+    }
+
+    function _checkIfExchangeAllowedForPair(
+        address tokenA,
+        address tokenB,
+        uint256 amount,
+        address trader
+    ) internal view returns (uint256 tokenAAmount, uint256 tokenBAmount) {
+        PriceThreshold memory tokenAThreshold = _thresholds[tokenA];
+        PriceThreshold memory tokenBThreshold = _thresholds[tokenB];
+
+        (uint256 tokenASellAmount, uint256 tokenBBuyAmount) = estimateTokenReturns(tokenA, tokenB, amount);
+        uint256 balanceOfAToken = ERC20(tokenA).balanceOf(trader) * (10 ** getDecimals(tokenA));
+        uint256 balanceOfBToken = ERC20(tokenB).balanceOf(address(this)) * (10 ** getDecimals(tokenB));
+
+        if (tokenASellAmount > balanceOfAToken) {
+            revert InvalidAmount("No enough tokens to sell");
+        } else if (tokenBBuyAmount > balanceOfBToken) {
+            revert InvalidAmount("No enough tokens to buy");
+        }
+
+        if (tokenAThreshold.maxSellAmount != 0 && tokenAThreshold.interval < block.timestamp) {
+            if ((_sellAmounts[tokenA] + tokenASellAmount) > tokenAThreshold.maxSellAmount) {
+                revert InvalidAmount("Max sell amount of tokens exceeded");
+            }
+        }
+
+        if (tokenBThreshold.maxBuyAmount != 0 && tokenBThreshold.interval < block.timestamp) {
+            if ((_buyAmounts[tokenB] + tokenBBuyAmount) > tokenBThreshold.maxBuyAmount) {
+                revert InvalidAmount("Max buy amount of tokens exceeded");
+            }
+        }
+
+        return (tokenASellAmount, tokenBBuyAmount);
+    }
+
+    function estimateTokenReturns(
+        address tokenA,
+        address tokenB,
+        uint256 amount
+    ) public view returns (uint256 tokenASellAmount, uint256 tokenBBuyAmount) {
+        if (_buyPrices[tokenB].interval == 0 || _sellPrices[tokenA].interval == 0) {
+            revert NoPriceExists("Sell or buy price for pair not found");
+        } else if (_buyPrices[tokenB].interval > block.timestamp || _sellPrices[tokenA].interval > block.timestamp) {
+            revert NoPriceExists("Sell or buy price for pair is not valid");
+        }
+
+        uint256 tokenADecimals = 10 ** getDecimals(tokenA);
+        uint256 tokenBDecimals = 10 ** getDecimals(tokenB);
+
+        if (tokenADecimals != tokenBDecimals) {
+            uint256 tokenASellAmount = amount * _sellPrices[tokenA].price;
+            uint256 tokenBBuyAmount = ((tokenASellAmount * _sellPrices[tokenA].price) / _buyPrices[tokenB].price);
+
+            return (tokenASellAmount * tokenADecimals, tokenBBuyAmount * tokenBDecimals);
+        } else {
+            uint256 tokenASellAmount = (amount * tokenADecimals) * _sellPrices[tokenA].price;
+            uint256 tokenBBuyAmount = ((tokenASellAmount * _sellPrices[tokenA].price) / _buyPrices[tokenB].price);
+
+            return (tokenASellAmount, tokenBBuyAmount);
+        }
+    }
+
+    function getDecimals(address token) internal view returns (uint8) {
```

Comment:
you don't need some custom method to get decimals, use IERC20Metadata from openzeppelin, specifically in your formula